### PR TITLE
#1644 Add button for VirtualizedTable Rows in Expression Builder

### PR DESCRIPTION
**Pull request by veenas1**
Fixes: #1644
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx [hunk 1]
```diff
@@ -537,7 +531,6 @@ VirtualizedTable.propTypes = {
 		PropTypes.func.isRequired,
 		PropTypes.number.isRequired
 	]),
-	onRowDoubleClick: PropTypes.func,
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx:540]: I think we should keep this.  It's optional so nothing should happen if we don't pass in `onRowDoubleClick` from the exression builder.

File: canvas_modules/harness/src/client/constants/json/functionlist.json [hunk 2]
```diff
@@ -6338,6 +6338,8 @@
 			{
 				"id": "globals",
 				"field_columns": {
+					"add_column_info": {
```
Comment (matthoward366) [inline canvas_modules/harness/src/client/constants/json/functionlist.json:6341]: Why is this needed? We shouldn't need to change the schema for this extra column.

File: canvas_modules/common-canvas/src/common-properties/controls/expression/expression-builder/expression-select-field-function.jsx [hunk 3]
```diff
@@ -265,7 +270,37 @@ export default class ExpressionSelectFieldOrFunction extends React.Component {
 		return contentObject;
 	}
 
+	createAddButtonContent(field, index) {
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/controls/expression/expression-builder/expression-select-field-function.jsx:273]: Do we need 2 functions here that do almost the exact same thing?  We could either pass in the function to use or pass in details so we know what function to use.

File: canvas_modules/common-canvas/src/common-properties/controls/expression/expression-builder/expression-select-field-function.jsx [hunk 4]
```diff
@@ -265,7 +270,37 @@ export default class ExpressionSelectFieldOrFunction extends React.Component {
...
+		const addButtonContent = (
+			<Button
+				className="expression-add-field-button properties-expr-table-cell"
+				onClick={() => (this.state.selectedIndex === 0 ? this.onAddFieldClick(index) : this.onAddFunctionClick(index))}
+				kind="ghost"
+				size="small"
+			>
+				<Add16 aria-label="Add" />
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/controls/expression/expression-builder/expression-select-field-function.jsx:281]: This needs to be translated.

**Review by matthoward366 — CHANGES_REQUESTED** [review]

File: canvas_modules/common-canvas/__tests__/test_resources/json/expression-function-list.json [hunk 5]
```diff
@@ -989,6 +989,8 @@
             "default": "Globals"
           },
           "field_columns": {
+            "add_column_info": {
```
Comment (matthoward366) [inline canvas_modules/common-canvas/__tests__/test_resources/json/expression-function-list.json:992]: I think these changes also need to be reverted?

File: canvas_modules/common-canvas/__tests__/common-properties/controls/expression-test.js [hunk 6]
```diff
@@ -127,6 +127,10 @@ function getCopy(value) {
 	return JSON.parse(JSON.stringify(value));
 }
 
+function getAddButtonsList(rows) {
+	return rows.find(".ReactVirtualized__Table__rowColumn").find("Button.expression-add-field-button");
```
Comment (matthoward366) [inline canvas_modules/common-canvas/__tests__/common-properties/controls/expression-test.js:131]: Why do we need `ReactVirtualized__Table__rowColumn`?  I'm asking because this might change in the future.  Wouldn't the button be unique enough?

**Review by matthoward366 — COMMENTED** [review]
Can we do something to make it so the add button row stays the same width?  In some case it's using too much space and caused an ellipsis.
Screenshot 2023-12-19 at 10 59 48 AM

Also, I found a defect when sorting a column then clicking on it.  The wrong row is selected.  This isn't caused by your change but can you have a look to see if it's an easy fix or create a new issue for it.

**Comment by veenas1** [comment]
@matthoward366 
- Add column width: Could you please check now, I have made the width to be in px, so that it's used as is and not adjusted. 
- Column Sort: Unable to replicate the issue.

File: canvas_modules/common-canvas/src/common-properties/controls/expression/expression-builder/expression-select-field-function.jsx [hunk 7]
```diff
@@ -377,17 +418,29 @@ export default class ExpressionSelectFieldOrFunction extends React.Component {
...
+					key: "fieldName",
+					label: categoryInfo.field_columns.field_column_info.locLabel,
+					description: categoryInfo.field_columns.field_column_info.descLabel,
+					resizable: true
+				});
+			valueHeader.push(
+				{
+					key: "addColumn",
+					label: formatMessage(this.reactIntl, MESSAGE_KEYS.EXPRESSION_ADD_COLUMN),
+					width: "50px",
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/controls/expression/expression-builder/expression-select-field-function.jsx:437]: For all the addColumn fields we should also add `addColumn` `staticWidth: true`

File: canvas_modules/common-canvas/src/common-properties/controls/expression/expression-builder/expression-select-field-function.jsx [hunk 8]
```diff
@@ -377,17 +418,29 @@ export default class ExpressionSelectFieldOrFunction extends React.Component {
...
+				},
+				{
+					key: "values",
+					label: categoryInfo.field_columns.value_column_info.locLabel,
+					description: categoryInfo.field_columns.value_column_info.descLabel,
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/controls/expression/expression-builder/expression-select-field-function.jsx:442]: Remove comma at the end

File: canvas_modules/common-canvas/src/common-properties/controls/expression/expression-builder/expression-select-field-function.jsx [hunk 9]
```diff
@@ -597,10 +651,12 @@ export default class ExpressionSelectFieldOrFunction extends React.Component {
 			MESSAGE_KEYS.EXPRESSION_FUNCTION_COLUMN);
 		const returnColumn = formatMessage(this.reactIntl,
 			MESSAGE_KEYS.EXPRESSION_RETURN_COLUMN);
+		const addNewColumn = formatMessage(this.reactIntl,
+			MESSAGE_KEYS.EXPRESSION_ADD_COLUMN);
 
-
-		headers.push({ key: "function", label: functionColumn, width: 73, resizable: true });
-		headers.push({ key: "return", label: returnColumn, width: 27 });
+		headers.push({ key: "addColumn", label: addNewColumn, width: "50px", resizable: true });
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/controls/expression/expression-builder/expression-select-field-function.jsx:657]: This shouldn't have resizable

**Comment by matthoward366** [comment]
Here is what I'm seeing for sort.  The selection gets off.  
sort issue

File: canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx
Comment (veenas1) [inline canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx:1]: @matthoward366 once sorted, the current row index to be considered.

File: canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx [hunk 10]
```diff
@@ -380,7 +380,7 @@ class VirtualizedTable extends React.Component {
 		}
 
 		if (this.props.selectable) {
-			const rowSelected = this.isRowSelected(rowData.originalRowIndex);
+			const rowSelected = this.props.sortDirection ? this.isRowSelected(rowData.index) : this.isRowSelected(rowData.originalRowIndex); // use current row index when Sorted
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx:383]: @caritaou Do you see any issue with this change?  I did some testing around row selection and for the most part it doesn't really work correctly onSort in the table controls.  It just clears the selection.

Comment (matthoward366) [reply canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx:383]: but this does fix the issue in the expressionBuilder around sorting and selection.

**Review by matthoward366 — APPROVED** [review]
lgtm.  @caritaou can you merge if you think the selection change is good.

File: canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx (hunk 10 above)
Comment (caritaou) [reply canvas_modules/common-canvas/src/common-properties/components/virtualized-table/virtualized-table.jsx:383]: This looks good to me, considering this never worked for sorting columns with row selections.
